modules/assim.sequential/R/Analysis_sda_block.R

```diff
@@ -632,4 +639,99 @@ block.2.vector <- function (block.list, X, H) {
               Pf = Pf,
               mu.a = mu.a,
               Pa = Pa))
+}
+
+##' This function provides means to split large SDA analysis (MCMC) runs into separate `qsub` jobs.
+##' Including job creation, submission, and assemble.
+##' @title qsub_analysis_submission
+##' @param block.list list: MCMC configuration lists for the block SDA analysis.
+##' @param outdir character: SDA output path.
+##' @param job.per.folder numeric: number of jobs per folder.
+##' @export
+##' 
+qsub_analysis_submission <- function(block.list, outdir, job.per.folder = 200) {
+  L <- length(block.list)
+  # calculate proper folder number based on settings.
+  folder.num <- ceiling(L/job.per.folder)
+  # create folder.
+  # if we have previous outputs, remove them.
+  if (file.exists(file.path(outdir, "qsub_analysis"))) {
+    unlink(file.path(outdir, "qsub_analysis"), recursive = T)
+  }
+  # create new folder.
+  dir.create(file.path(outdir, "qsub_analysis"))
+  # loop over sub-folders.
+  folder.paths <- c()
+  PEcAn.logger::logger.info(paste("Submitting", folder.num, "jobs."))
+  for (i in 1:folder.num) {
+    # create folder for each set of job runs.
+    # calculate start and end index for the current folder.
+    head.num <- (i-1)*job.per.folder + 1
+    if (i*job.per.folder > L) {
+      tail.num <- L
+    } else {
+      tail.num <- i*job.per.folder
+    }
+    # naming and creating folder.
+    folder.name <- paste0("From_", head.num, "_to_", tail.num)
+    folder.path <- file.path(outdir, "qsub_analysis", folder.name)
+    folder.paths <- c(folder.paths, folder.path)
+    dir.create(folder.path)
+    # save corresponding block list to the folder.
+    blocks <- block.list[head.num:tail.num]
+    save(blocks, file = file.path(folder.path, "block.Rdata"))
+    # create job file.
+    jobsh <- readLines(con = system.file("analysis_qsub.job", package = "PEcAn.ModelName"), n = -1, warn=FALSE)
+    jobsh <- gsub("@FOLDER_PATH@", folder.path, jobsh)
+    writeLines(jobsh, con = file.path(folder.path, "job.sh"))
+    # qsub command.
+    qsub <- "qsub -l h_rt=48:00:00 -l buyin -pe omp 28 -V -N @NAME@ -o @STDOUT@ -e @STDERR@ -S /bin/bash"
+    qsub <- gsub("@NAME@", paste0("Job-", i), qsub)
+    qsub <- gsub("@STDOUT@", file.path(folder.path, "stdout.log"), qsub)
+    qsub <- gsub("@STDERR@", file.path(folder.path, "stderr.log"), qsub)
+    qsub <- strsplit(qsub, " (?=([^\"']*\"[^\"']*\")*[^\"']*$)", perl = TRUE)
+    cmd <- qsub[[1]]
+    out <- system2(cmd, file.path(folder.path, "job.sh"), stdout = TRUE, stderr = TRUE)
+  }
+  # checking results.
+  PEcAn.logger::logger.info("Checking results.")
+  while (length(list.files(outdir, pattern="results.Rdata", recursive = T)) < folder.num) {
+    Sys.sleep(60)
+  }
```

Comment:
Flagging that this has high potential to become an infinite loop if any job fails / is rejected by the queue / etc

Reply:
More generally, I'm cautious of submitting anything to qsub and waiting for it to complete before returning from the submitting function -- Part of the point of qsub is that it does the waiting for you.

Reply:
I am not sure if I resolve this correctly. Here, instead, I detect the job completion by checking if the job still exists on the server. This will ensure a finite `while` loop even if, for some reason, some jobs never get finished and are killed by the server.

Reply:
> detect the job completion by checking if the job still exists on the server

👍 That seems like a solid improvement. -- still has the potential to be waiting a long time if the queue is stalled / moving slowly, but will only be waiting for outputs the server still plans to provide.